File: python/tools/registry.py
# Copyright since 2025 Mifos Initiative
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import re
from typing import Any, List

from tools.domains.accounting import create_journal_entry, get_journal_entries, list_gl_accounts
from tools.domains.charges import create_charge, get_charge, list_charges, update_charge
from tools.domains.clients import (
    activate_client,
    apply_client_charge,
    close_client,
    create_client,
    create_client_identifier,
    create_group,
    get_client_accounts,
    get_client_addresses,
    get_client_charges,
    get_client_details,
    get_client_documents,
    get_client_identifiers,
    get_client_transactions,
    get_group_details,
    search_clients_by_name,
    update_client_mobile,
)
from tools.domains.codetables import get_code_values, list_codes, list_datatables
from tools.domains.groups import (
    activate_group,
    add_group_member,
    create_center,
    get_center,
    list_centers,
    list_groups,
)
from tools.domains.loans import (
    apply_late_fee,
    approve_and_disburse_loan,
    create_group_loan,
    create_loan,
    get_loan_details,
    get_loan_template,
    get_overdue_loans,
    get_repayment_schedule,
    make_loan_repayment,
    reject_loan_application,
    reschedule_loan,
    undo_loan_approval,
    undo_loan_disbursal,
    waive_interest,
)
from tools.domains.products import (
    get_loan_product,
    get_savings_product,
    list_loan_products,
    list_savings_products,
)
from tools.domains.reports import create_report, get_report, list_reports, run_report, update_report
from tools.domains.savings import (
    apply_savings_charge,
    approve_and_activate_savings,
    calculate_and_post_interest,
    close_savings_account,
    create_savings_account,
    deposit_savings,
    get_savings_account,
    get_savings_transactions,
    withdraw_savings,
)
from tools.domains.staff import get_office_details, get_staff_details, list_offices, list_staff

# ✅ AI Suggestion Engine
from core.suggestion_engine import generate_suggestions


class DomainRegistry:
    """
    Domain router that maps user intent to a filtered subset of MCP tools.
    """

    def __init__(self):
        self.domain_map = {
            "clients": [
                search_clients_by_name, get_client_details, get_client_accounts,
                create_client, activate_client, update_client_mobile, close_client,
                create_group, get_group_details, get_client_identifiers,
                create_client_identifier, get_client_documents, get_client_charges,
                apply_client_charge, get_client_transactions, get_client_addresses
            ],
            "groups": [
                list_groups, activate_group, add_group_member,
                list_centers, get_center, create_center
            ],
            "loans": [
                get_loan_details, get_repayment_schedule, create_loan,
                approve_and_disburse_loan, reject_loan_application,
                make_loan_repayment, apply_late_fee, waive_interest,
                get_overdue_loans, create_group_loan,
                undo_loan_approval, undo_loan_disbursal,
                get_loan_template, reschedule_loan
            ],
            "savings": [
                get_savings_account, get_savings_transactions, create_savings_account,
                approve_and_activate_savings, close_savings_account, deposit_savings,
                withdraw_savings, apply_savings_charge, calculate_and_post_interest
            ],
            "staff": [
                list_staff, get_staff_details, list_offices, get_office_details
            ],
            "accounting": [
                list_gl_accounts, get_journal_entries, create_journal_entry
            ],
            "reports": [
                list_reports, get_report, run_report, create_report, update_report
            ],
            "products": [
                list_loan_products, get_loan_product,
                list_savings_products, get_savings_product
            ],
            "charges": [
                list_charges, get_charge, create_charge, update_charge
            ],
            "codetables": [
                list_codes, get_code_values, list_datatables
            ],
        }

    def get_domain(self, domain: str) -> List[Any]:
        return self.domain_map.get(domain, [])

    def get_all_tools(self) -> List[Any]:
        all_tools = []
        seen = set()

        for tools in self.domain_map.values():
            for tool in tools:
                if tool.name not in seen:
                    all_tools.append(tool)
                    seen.add(tool.name)

        return all_tools

   def route_intent(self, user_query: str) -> List[Any]:
    """Route user query to relevant domain tools using word-boundary matching."""
    query = user_query.lower()
    active_domains = set()

    def matches_keyword(keywords: List[str]) -> bool:
        for word in keywords:
            pattern = rf"\b{re.escape(word)}\b"
            if re.search(pattern, query):
                return True
        return False

    # Loans
    if matches_keyword([
        "loan", "repayment", "disburse", "overdue", "arrear",
        "reject", "waive", "installment", "undo", "reschedule", "template"
    ]):
        active_domains.add("loans")

    # Clients
    if matches_keyword([
        "client", "person", "search", "activate", "mobile", "kyc",
        "identifier", "address", "charge", "fee", "document"
    ]):
        active_domains.add("clients")

    # Groups
    if matches_keyword([
        "group", "center", "centre", "member"
    ]):
        active_domains.add("groups")

    # Savings
    if matches_keyword([
        "saving", "deposit", "withdraw", "balance", "wallet", "interest"
    ]):
        active_domains.add("savings")

    # Staff
    if matches_keyword([
        "staff", "officer", "employee", "office", "branch"
    ]):
        active_domains.add("staff")

    # Accounting
    if matches_keyword([
        "journal", "ledger", "account", "debit", "credit", "accounting"
    ]):
        active_domains.add("accounting")

    # Reports
    if matches_keyword([
        "report", "generate", "sql"
    ]):
        active_domains.add("reports")

    # Products
    if matches_keyword([
        "product"
    ]):
        active_domains.add("products")

    # Charges
    if matches_keyword([
        "charge", "penalty", "fee"
    ]):
        active_domains.add("charges")

    # Code tables
    if matches_keyword([
        "code", "datatable", "dropdown"
    ]):
        active_domains.add("codetables")

    # Default fallback
    if not active_domains:
        active_domains.add("clients")

    # Deduplicate tools
    seen = set()
    result = []

    for domain in active_domains:
        for tool in self.domain_map[domain]:
            if tool.name not in seen:
                result.append(tool)
                seen.add(tool.name)

    return result

    def get_suggestions(self, intent: str, data: Any) -> List[str]:
        return generate_suggestions(intent, data)


# Global router instance
router = DomainRegistry()

File: rust/src/registry.rs
// Copyright since 2025 Mifos Initiative
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

use std::collections::{HashMap, HashSet};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use crate::embeddings::SemanticRouter;

/// Minimum similarity score required for a domain to be considered relevant.
pub const INTENT_ROUTE_THRESHOLD: f32 = 0.20;

/// Request schema for the Hybrid Intent Router tool.
#[derive(Debug, Serialize, Deserialize, JsonSchema)]
pub struct IntentRouterReq {
    /// The natural language banking query to resolve into relevant tools.
    /// Example: "I need to check if a client has any overdue installments"
    pub query: String,
}

/// DomainRegistry provides NLP intent routing for Bank Staff queries.
/// This is used by LLM agents (e.g., LangGraph) that wish to aggressively prune the Context Window.
///
/// Supports two routing modes:
/// - **Keyword** (`route_intent`): Zero-latency string matching (~3us). Always available.
/// - **Semantic** (`route_intent_semantic`): ML-powered cosine similarity (~5-10ms). Requires model init.
#[derive(Clone, Debug)]
pub struct DomainRegistry {
    pub domain_map: HashMap<&'static str, Vec<&'static str>>,
    pub semantic_router: Option<SemanticRouter>,
}

/// Builds the canonical domain -> tool mapping.
/// This is the single source of truth — every tool registered in server.rs must appear here.
fn build_domain_map() -> HashMap<&'static str, Vec<&'static str>> {
    let mut map = HashMap::new();

    map.insert("clients", vec![
        "search_clients_by_name", "get_client_details", "get_client_accounts",
        "create_client", "activate_client", "update_client_mobile", "update_client",
        "delete_client", "close_client",
        "get_client_identifiers", "create_client_identifier",
        "get_client_documents", "get_client_charges",
        "apply_client_charge", "pay_client_charge", "waive_client_charge",
        "get_client_transactions", "get_client_transaction",
        "undo_client_transaction", "get_client_addresses",
        "list_client_collaterals", "get_client_collateral",
        "create_client_collateral", "update_client_collateral", "delete_client_collateral",
    ]);

    map.insert("groups", vec![
        "list_groups", "get_group", "create_group",
        "activate_group", "add_group_member", "remove_group_member",
        "get_group_accounts", "create_group_savings_account",
        "update_group", "close_group",
        "list_centers", "get_center", "create_center",
    ]);

    map.insert("loans", vec![
        "get_loan_details", "get_repayment_schedule", "get_loan_history",
        "get_overdue_loans",
        "create_loan", "update_loan", "delete_loan",
        "list_loan_products", "create_group_loan",
        "approve_and_disburse_loan", "reject_loan_application",
        "make_loan_repayment", "apply_late_fee", "waive_interest",
    ]);

    map.insert("collaterals", vec![
        "list_loan_collaterals", "get_loan_collateral", "create_loan_collateral",
        "update_loan_collateral", "delete_loan_collateral",
    ]);

    map.insert("savings", vec![
        "get_savings_account", "get_savings_transactions", "create_savings_account",
        "list_savings_products",
        "approve_and_activate_savings", "close_savings_account",
        "deposit_savings", "withdraw_savings",
        "apply_savings_charge", "calculate_and_post_interest",
    ]);

    map.insert("staff", vec![
        "list_staff", "get_staff_details", "list_offices", "get_office_details",
    ]);

    map.insert("accounting", vec![
        "list_gl_accounts", "get_journal_entries", "create_journal_entry",
    ]);

    map.insert("bulk", vec![
        "bulk_search_clients", "bulk_get_loan_status", "bulk_disburse_loans",
        "bulk_make_repayments", "bulk_activate_clients", "bulk_get_savings_balances",
        "bulk_apply_fees", "bulk_close_accounts", "bulk_create_savings_accounts",
        "bulk_approve_and_activate_savings", "bulk_deposit_savings",
    ]);

    map.insert("charges", vec![
        "retrieve_charge", "create_charge", "update_charge", "delete_charge",
    ]);

    map
}

impl DomainRegistry {
    /// Full initialization: builds domain map + starts background ML loading.
    pub fn new() -> Self {
        let domain_map = build_domain_map();
        let semantic_router = Some(SemanticRouter::new());

        Self { domain_map, semantic_router }
    }

    /// Create a registry without the semantic router (for benchmarks and tests).
    pub fn new_keywords_only() -> Self {
        Self {
            domain_map: build_domain_map(),
            semantic_router: None,
        }
    }

    pub fn get_domain(&self, domain: &str) -> Vec<&'static str> {
        self.domain_map.get(domain).cloned().unwrap_or_default()
    }

    pub fn get_all_tools(&self) -> Vec<&'static str> {
        let mut tools = Vec::new();
        let mut seen = HashSet::new();
        for list in self.domain_map.values() {
            for t in list {
                if seen.insert(*t) {
                    tools.push(*t);
                }
            }
        }
        tools
    }

    /// Internal raw keyword matcher. 
    /// Returns an empty Vec if no matches are found (no fallback).
    fn route_keywords_internal(&self, query: &str) -> Vec<&'static str> {
        let mut active_domains = HashSet::new();

        if ["loan", "repayment", "disburse", "overdue", "arrear", "reject", "waive", "installment", "product", "interest", "principal", "emi", "npa", "write-off", "write off", "reschedule", "prepay", "payoff", "amortiz", "moratorium", "grace period"].iter().any(|w| query.contains(w)) {
            active_domains.insert("loans");
        }
        if ["client", "person", "customer", "borrower", "applicant", "search", "activate", "mobile", "kyc", "identifier", "address", "document", "onboard", "enroll", "register"].iter().any(|w| query.contains(w)) {
            active_domains.insert("clients");
        }
        if ["collateral", "security", "pledge", "lien", "asset", "guarantee", "mortgage"].iter().any(|w| query.contains(w)) {
            active_domains.insert("collaterals");
        }
        if ["group", "center", "centre", "member", "lending group", "solidarity", "village banking"].iter().any(|w| query.contains(w)) {
            active_domains.insert("groups");
        }
        if ["saving", "deposit", "withdraw", "balance", "wallet", "fixed deposit", "recurring", "fd ", "rd "].iter().any(|w| query.contains(w)) {
            active_domains.insert("savings");
        }
        if ["staff", "officer", "employee", "office", "branch", "manager", "teller", "cashier", "supervisor"].iter().any(|w| query.contains(w)) {
            active_domains.insert("staff");
        }
        if ["journal", "ledger", "gl account", "gl ", "debit", "credit", "accounting", "general ledger", "chart of accounts", "reconcil", "trial balance", "posting"].iter().any(|w| query.contains(w)) {
            active_domains.insert("accounting");
        }
        if ["bulk", "multiple", "batch", "parallel", "concurrent", "all at once", "mass", "each", "both", "all of them", "everyone", "all clients", "all loans", "all savings"].iter().any(|w| query.contains(w)) {
            active_domains.insert("bulk");
        }
        if ["charge", "fee", "penalty", "tax", "fee definition", "surcharge", "levy", "service charge", "processing fee"].iter().any(|w| query.contains(w)) {
            active_domains.insert("charges");
        }

        let mut result = Vec::new();
        let mut seen = HashSet::new();
        for d in active_domains {
            if let Some(tools) = self.domain_map.get(d) {
                for t in tools {
                    if seen.insert(*t) {
                        result.push(*t);
                    }
                }
            }
        }
        result
    }

    /// Keyword-based intent routing. Returns "clients" tools by default if no matches.
    pub fn route_intent(&self, user_query: &str) -> Vec<&'static str> {
        let query = user_query.to_lowercase();
        let kw_tools = self.route_keywords_internal(&query);
        
        if kw_tools.is_empty() {
            self.get_domain("clients")
        } else {
            kw_tools
        }
    }

    /// Hybrid intent routing: Optimized for speed and accuracy.
    ///
    /// 1. **Fast Path** (Keyword): Returns immediately if keywords are found.
    /// 2. **Slow Path** (Semantic Rescue): Only triggers if keywords fail (0 matches).
    pub async fn route_intent_hybrid(&self, user_query: &str, threshold: f32) -> Vec<&'static str> {
        let query = user_query.to_lowercase();
        
        // 1. Fast Path: Keyword Match
        let kw_tools = self.route_keywords_internal(&query);
        if !kw_tools.is_empty() {
            return kw_tools;
        }

        // 2. Slow Path: Semantic Rescue (only if keywords failed)
        if let Some(ref router) = self.semantic_router {
            match router.route(user_query, threshold).await {
                Ok(scored_domains) if !scored_domains.is_empty() => {
                    let mut result = Vec::new();
                    let mut seen = HashSet::new();
                    for (domain_name, _score) in scored_domains {
                        if let Some(tools) = self.domain_map.get(domain_name) {
                            for t in tools {
                                if seen.insert(*t) {
                                    result.push(*t);
                                }
                            }
                        }
                    }
                    if !result.is_empty() {
                        return result;
                    }
                }
                _ => {}
            }
        }

        // Final Fallback: Default to "clients"
        self.get_domain("clients")
    }
}
